### ISSUE TITLE
Inconsistent results using Libvirt + swtpm. Error: Need read rights on issuer certificate

### ISSUE THREAD

**Issue by nmagnezi** [stefanberger/swtpm#644]
**Describe the bug**
Hello. I'm a developer in a project named [assisted-service](
https://github.com/openshift/assisted-service), and we use swtpm as a part of our CI.

We use CentOS 8 based hypervisors with Libvirt with swtpm to emulate nodes we deploy OpenShift (Red Hat enterprise Kubernetes). We usually use 5 VMs per CI Job. Libvirt configuration takes place with our [testing framework](https://github.com/openshift/assisted-test-infra) that [uses terraform](https://github.com/openshift/assisted-test-infra/blob/2357f4b3d175b522da27e5aa74a7459642ff111e/terraform_files/baremetal/variables-libvirt.tf#L191-L201), for that matter.

I noticed that swtpm sometimes fail with this error:
`Need read rights on issuer certificate /var/lib/swtpm-localca/issuercert.pem for user tss.`

This is odd since it is inconsistent (it does not always fail), and other VMs that reference the same cert gets configured with no issues. I'll attach all logs below.

I have made sure we use a recent swtpm version and even [attempted to use Libvirt-7.x instead](https://github.com/openshift/assisted-test-infra/pull/1406#issuecomment-1025187922) of Libvirt-6.x, yet I got the same result. 

**To Reproduce**
1. Reproduced in CI with Libvirt + swtpm, as detailed above.

**Expected behavior**
I'd expect consistent results. swtpm configurations should succeed in the case I mentioned above. They all use the same configuration with the same cert file.

**Desktop (please complete the following information):**
 - OS: CentOS
 - Version 8

**Versions of relevant components**
- swtpm x86_640.5.2-3.20201226gite59c0c1.el8      
- libtpms x86_640.8.6-0.20210910git7a4d46a119.el8.0
- openssl x86_64  1:1.1.1k-5.el8_5
- gnutls  x86_64   3.3.29-9.el7_6
- swtpm-tools x86_640.5.2-3.20201226gite59c0c1.el8

**Log files**
Find the CI jobs logs: [here](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-assisted-test-infra-master-e2e-metal-assisted-tpmv2-periodic/1485803276946378752#1:build-log.txt%3A9194).
CI swtpm Logs: [here](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-assisted-test-infra-master-e2e-metal-assisted-tpmv2-periodic/1485803276946378752/artifacts/e2e-metal-assisted-tpmv2-periodic/baremetalds-assisted-gather/artifacts/libvirt-qemu/)

I'll also upload them to gist in case the CI logs will go away:
[CI Log](https://gist.github.com/nmagnezi/0a5080a9e052fc6f29a90e126c2039b6)
[ ls-swtpm-localca.txt](https://gist.github.com/nmagnezi/ebbb6e8256c351141ffc97ddb9586815)
[test-infra-cluster-ccedb7e5-master-0-swtpm.log](https://gist.github.com/nmagnezi/c52ea376374e4076215b93bc33e9c6da) --> Success
 [test-infra-cluster-ccedb7e5-master-1-swtpm.log](https://gist.github.com/nmagnezi/c11e845a119d38d7c334a749fecc61c2) --> Success
[test-infra-cluster-ccedb7e5-master-2-swtpm.log](https://gist.github.com/nmagnezi/947772dff4e70a22016669776c366afa) --> Error
[test-infra-cluster-ccedb7e5-worker-0-swtpm.log](https://gist.github.com/nmagnezi/247fa2f9a8bbde5b35be8e48a1a3cf11) --> Success
[test-infra-cluster-ccedb7e5-worker-1-swtpm.log](https://gist.github.com/nmagnezi/19464650dc7f93f659f2cfa70d1b91ca) --> Success

**Additional context**
As you'll [see here](
https://prow.ci.openshift.org/job-history/gs/origin-ci-test/logs/periodic-ci-openshift-assisted-test-infra-master-e2e-metal-assisted-tpmv2-periodic), this does not always fail, which makes it harder to debug. I'd appreciate your advice.

**Comment by stefanberger** [comment]
Are you starting multiple VMs concurrently? There would be a possibility that I could see that some certificate authority files are not available, yet. Upon the first start of a VM with an attached vTPM the certificate authority is created, unless one has been setup beforehand, and the possibility there is that while one process creates the CA, the others are already trying to create VMs while the CA is not fully there yet.

**Comment by stefanberger** [comment]
PR #645 should resolve the issue. Can you try this in your environment?

**Comment by stefanberger** [comment]
Other PRs for older branches: #646 #647

**Comment by osherdp** [comment]
@stefanberger sure thing! we'll give it a couple of runs and see if it fixes it
yes, we use https://github.com/dmacvicar/terraform-provider-libvirt that will surely sometimes create the VMs at the same time, so it's very likely this is the issue we're having
Again, thank you for your quick response! and we'll come back with an answers

**Comment by stefanberger** [comment]
Yes, please, let me know.

**Comment by nmagnezi** [comment]
> PR #645 should resolve the issue. Can you try this in your environment?

@stefanberger thanks so much for the prompt response.
As @osherdp said, we do start multiple VMs concurrently.
We currently deploy swtpm from an RPM, but yes we will try to deploy it with the fix you mentioned and report results back here.

**Comment by nmagnezi** [comment]
Hi @stefanberger, sorry it took some time, but I ran into an issue while trying to use [PR-645](https://github.com/stefanberger/swtpm/pull/645) in our CI.

What I did is to[ fetch your branch and install](
https://github.com/openshift/assisted-test-infra/pull/1433/files#diff-e1bb14592e615972cef412920fc46d0ed0061caa78439a08c3165c51cdba585aR59-R65), instead of deploying from an RPM, but [based on the logs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_assisted-test-infra/1433/pull-ci-openshift-assisted-test-infra-master-e2e-metal-assisted-tpmv2/1491433490523099136/artifacts/e2e-metal-assisted-tpmv2/baremetalds-assisted-gather/artifacts/libvirt-qemu/test-infra-cluster-a165265c-master-0-swtpm.log), it is missing swtpm_cert:
```bash
Could not find swtpm_cert in PATH.
swtpm_localca exit with status 1: 
```

I tried to follow the [RPM spec](https://src.fedoraproject.org/rpms/swtpm/blob/f35/f/swtpm.spec) to get it generated with no success. Can you tell me how to generate it?

**Comment by stefanberger** [comment]
What does your `./configure` line print out? You need the gnutls packages gnutls, gnutls-devel and gnutls-utils installed to get this here:

```
[...]
config.status: config.h is unchanged
config.status: executing libtool commands
config.status: executing depfiles commands

with_gnutls     :   yes  (no = swtpm_cert will NOT be built)   <------------------------------
with_selinux    :    no  (no = SELinux policy extensions will NOT be built)
```

**Comment by nmagnezi** [comment]
I've updated the steps to [this](https://github.com/openshift/assisted-test-infra/pull/1433/files#diff-e1bb14592e615972cef412920fc46d0ed0061caa78439a08c3165c51cdba585aR59-R67), locally now it does print:
```bash
with_gnutls     :   yes  (no = swtpm_cert will NOT be built)
with_selinux    :    no  (no = SELinux policy extensions will NOT be built)
with_cuse       :    no  (no = no CUSE interface)
with_chardev    :   yes  (no = no chardev interface)
with_vtpm_proxy :   yes  (no = no vtpm proxy support; Linux only)
with_seccomp    :   yes  (no = no seccomp profile; Linux only)
```

Will update with the CI results. thanks again!

**Comment by nmagnezi** [comment]
Hi @stefanberger,
We had multiple successful runs with https://github.com/stefanberger/swtpm/pull/645 in our CI.
I planned to give this a few more runs (due to the nature of inconsistent results this issue caused), but our CI currently face some unrelated infrastructure issue.

I wanted to keep you posted that it looks promising, and once I get some more results I'll report back here. Thanks again.

**Comment by stefanberger** [comment]
Please let me know how it's going. I intend to create fix-releases of swtpm soon also due to another necessary fix.

**Comment by stefanberger** [comment]
Uuuh. I had been trying to recreate the issue but had not been successful doing it until now.... I have tried it with this command line that removes the signing key and issuer certificate and creates 100 concurrent swtpm_setup runs where the first one needs to create the signing key in a 1st step and then in the 2nd step it creates the issue cert. Other processes will then either wait in the file lock or, if they found the signing key, should proceed to creating the cert. 

```
 rm -rf /var/lib/swtpm-localca/* ; for ((i = 0; i<100;i++)); do rm -rf /tmp/vtpm${i}; mkdir -p /tmp/vtpm${i}; swtpm_setup --tpm2 --tpmstate /tmp/vtpm${i} --create-ek-cert --create-platform-cert --pcr-banks sha256 2>&1 | grep -i read & done
```
[my machine has 16 CPU threads]
The above is grepping for the 'read' in 'Need read rights'.

The problem with it was they probably all ended in the file lock. So, I now increased it to **200 concurrent swtpm_setup** runs and I do see the error message now as well. Some processes seem to find the signing key but then the issuercert is not there, yet. Fix for master and stable-0.7 are good.... will merge them by tomorrow.
